lib/services/plugins-service.ts

```diff
@@ -0,0 +1,223 @@
+		return path.join(this.nodeModulesPath, moduleName, "package.json");
+	}
+	
+	private getDependencies(): string[] {
+		let packageJsonFilePath = this.getPackageJsonFilePath();
+		return _.keys(require(packageJsonFilePath).dependencies);
+	}
+	
+	private getNodeModuleData(moduleName: string): IFuture<INodeModuleData> {
+		return (() => { 
+			let packageJsonFilePath = this.getPackageJsonFilePathForModule(moduleName);
+			if(this.$fs.exists(packageJsonFilePath).wait()) {
+				let data = require(packageJsonFilePath);
+				return {
+					name: data.name,
+					version: data.version,
+					fullPath: path.dirname(packageJsonFilePath),
+					isPlugin: data.nativescript !== undefined,
+					moduleInfo: data.nativescript			
+				};
+			}
+			
+			return null;
+		}).future<INodeModuleData>()();
+	}
+	
+	private convertToPluginData(cacheData: ICacheData): IPluginData {
+		let pluginData: any = {};
+		pluginData.name = cacheData.name;
+		pluginData.version = cacheData.version;
+		pluginData.fullPath = path.dirname(this.getPackageJsonFilePathForModule(cacheData.name));
+		pluginData.isPlugin = !!cacheData.nativescript;
+		
+		if(pluginData.isPlugin) {
+			pluginData.platformsData = cacheData.nativescript.platforms;
+		}
+		
+		return pluginData;
+	}
+	
+	private getAllInstalledModules(): IFuture<INodeModuleData[]> {
+		return (() => {
+			this.ensureAllDependenciesAreInstalled().wait();
+			this.$fs.ensureDirectoryExists(this.nodeModulesPath).wait();
+			
+			let nodeModules = this.getDependencies();
+			return _.map(nodeModules, nodeModuleName => this.getNodeModuleData(nodeModuleName).wait());
+		}).future<INodeModuleData[]>()();
+	}
+	
+	private executeNpmCommand(npmCommandName: string, npmCommandArguments: string): IFuture<string> {
+		return (() => {
+			let result = "";
+			
+			if(npmCommandName === PluginsService.INSTALL_COMMAND_NAME) {
+				result = this.$npm.install(npmCommandArguments, this.$projectData.projectDir, PluginsService.NPM_CONFIG).wait();				
+			} else if(npmCommandName === PluginsService.UNINSTALL_COMMAND_NAME) {
+				result = this.$npm.uninstall(npmCommandArguments, PluginsService.NPM_CONFIG).wait();
+			}
+			
+			return this.parseNpmCommandResult(result);
+		}).future<string>()();
+	}
+	
+	private parseNpmCommandResult(npmCommandResult: string): string {  // [[name@version, node_modules/name]]
+		return npmCommandResult[0][0].split("@")[0]; // returns plugin name
+	}
+	
+	private executeForAllInstalledPlatforms(action: (pluginDestinationPath: string, platform: string, platformData: IPlatformData) => void): void {
+		let availablePlatforms = _.keys(this.$platformsData.availablePlatforms);
+		_.each(availablePlatforms, platform => {
+			let isPlatformInstalled = this.$fs.exists(path.join(this.$projectData.platformsDir, platform.toLowerCase())).wait();
+			if(isPlatformInstalled) {
+				let platformData = this.$platformsData.getPlatformData(platform.toLowerCase());
```

Comment:
ahh, no, there's duplication of method's action paramater and each's argument - both are platform, which led me to this confusion. Consider renaming each's platform to availablePlatform or avlblPlatform or smth else.